### PR TITLE
Define sumcheck transcript configuration and Rust wrappers

### PR DESCRIPTION
**Pull request by yshekel**
- [x] C++ sumcheck transcript config + test
- [x]  Rust sumcheck transcript config + test
- [x] C++ sumcheck API and ffi
- [x] Rust sumcheck API and test

### CODE REVIEW

File: icicle/tests/test_hash_api.cpp [hunk 1]
```diff
@@ -1067,3 +1067,51 @@ TEST_F(HashApiTest, poseidon_tree)
 }
 
 #endif // POSEIDON
+
+/*======================= (TODO) Move to Sumcheck test-suite =======================*/
```
Comment (mickeyasa) [inline icicle/tests/test_hash_api.cpp:1071]: why didn't you open a new file for that?

File: icicle/tests/test_hash_api.cpp [hunk 2]
```diff
@@ -1067,3 +1067,51 @@ TEST_F(HashApiTest, poseidon_tree)
...
+{
+public:
+  // Helper function to convert a byte vector to a string for comparison
+  static std::string bytes_to_string(const std::vector<std::byte>& bytes)
+  {
+    return std::string(reinterpret_cast<const char*>(bytes.data()), bytes.size());
+  }
+};
+
+TEST_F(SumcheckTest, InitializeWithConstChar)
```
Comment (mickeyasa) [inline icicle/tests/test_hash_api.cpp:1084]: test name should be SumcheckTranscriptConfigInitializeWithConstChar
for filtering

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+   * - Endianness for encoding messages (little-endian by default).
+   * - The hashing function used to derive randomness.
+   * - A seed for initializing the random number generator (RNG).
+   *
+   * @tparam S Type of the field element (e.g., prime field or extension field elements).
+   */
+  template <typename S>
+  class SumcheckTranscriptConfig
+  {
+  private:
```
Comment (mickeyasa) [inline icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:22]: I prefer placing the private section at the bottom

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    /**
+     * @brief Converts a C-string to a vector of bytes.
+     *
+     * @param str Null-terminated C-string.
+     * @return std::vector<std::byte> Byte representation of the string.
+     */
+    static inline std::vector<std::byte> cstr_to_bytes(const char* str)
+    {
+      if (str == nullptr) return {};
+      size_t length = std::strlen(str);
```
Comment (mickeyasa) [inline icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:39]: const

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  };
+
+  /*
+ * Notes:
+ * 1. Both Prover and Verifier use this configuration to ensure deterministic randomness generation.
+ * 2. Messages (e.g., field elements, strings) are encoded in little/big-endian format based on
+ `SumcheckTranscriptConfig.little_endian`.
+
+ *
+ * Protocol Workflow:
```
Comment (mickeyasa) [inline icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:98]: this description is important and can move to the top

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * (1) Initialize Transcript:
+ * - entry_DS = [domain_separator_label || proof.num_vars || proof.degree || public (hardcoded?) || claimed_sum]
```
Comment (mickeyasa) [inline icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:101]: what is public (hardcoded?)

Comment (krakhit) [reply icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:102]: Public is any public data in the protocol that is known to both prover and verifier. User should provide this data , not sure the hardcoded was a correct terminology

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * (2) First Round:
+ * - entry_0 = [round_poly_label || r_0[x].len() || k=0 || r_0[x]]
+ * - alpha_0 = Hash(entry_DS || seed_rng || round_challenge_label || entry_0).to_field()
+ *
+ * (3) Round i > 0:
+ * - entry_i = [round_poly_label || r_i[x].len() || k=i || r_i[x]]
+ * - alpha_i = Hash(entry_0 || alpha_(i-1) || round_challenge_label || entry_i).to_field()
```
[the remark this answers is not in the document]
Comment (krakhit) [reply icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:110]: in every hash, we also need to include the entry_0 which contains the public data, prev hash output is in alpha_{I-1} 

This including public in every challenge is called strong Fiat Shamir, and it is debated whether weak Fiat Shamir (only including it in the first hash and then dropping it for the rest) is as secure. I felt it was best to go for strong  and include the public data encoded in entry_0 for all challenges.

**Review by krakhit — APPROVED** [review]
Looks good as per definitions. Added Suyash for a quick check.

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h (hunk 5 above)
Comment (krakhit) [reply icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:99]: Agreed

File: icicle/tests/test_hash_api.cpp (hunk 1 above)
Comment (yshekel) [reply icicle/tests/test_hash_api.cpp:1071]: I should do it yes. We should add a SUMCHECK feature to cmake right? and build the test accordingly

File: icicle/tests/test_hash_api.cpp (hunk 2 above)
Comment (yshekel) [reply icicle/tests/test_hash_api.cpp:1084]: your name is too short but ok I will rename. Although if I move it to a sumcheck test suite then I can remove the Sumcehck prefix from the test name I guess.

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h (hunk 3 above)
Comment (yshekel) [reply icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:22]: ok I can do that

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h (hunk 5 above)
Comment (yshekel) [reply icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:99]: will do

**Comment by yshekel** [comment]
> where are the rust/go bindings?

Rust in progress. Go is missing yes. I am not sure when will be done.

File: icicle/tests/test_hash_api.cpp (hunk 2 above)
Comment (mickeyasa) [reply icicle/tests/test_hash_api.cpp:1084]: do you plan to have a different test suite for FRI?
If so then I agree

Comment (yshekel) [reply icicle/tests/test_hash_api.cpp:1084]: I think so yes

File: icicle/include/icicle/sumcheck/sumcheck_transcript_config.h (hunk 6 above)
Comment (yshekel) [reply icicle/include/icicle/sumcheck/sumcheck_transcript_config.h:102]: @krakhit how should the user provide it? should I add it to the struct?

Comment (krakhit) [reply icicle/include/icicle/sumcheck/sumcheck_transcript_config.h]: yes, please include it in the transcript struct 
https://github.com/zkcrypto/merlin/blob/a94892f457202399661573480bb7250b4fe7c0af/src/transcript.rs#L245

**Pull request merged**